apps/pipelines/tests/test_runnable_builder.py

```diff
@@ -1220,3 +1222,38 @@ def _tool_call(route):
     assert output["messages"][-1] == "D d"
     output = runnable.invoke(PipelineState(messages=["z"], experiment_session=experiment_session))
     assert output["messages"][-1] == "A z"
+
+
+@pytest.mark.django_db()
+def test_router_node_output_structure(provider, provider_model, pipeline, experiment_session):
```

Comment:
Would you mind adding tests for the new functions as well, with happy and sad cases?

A suggestion:
I'm envisioning tests where the pipeline's state is defined in raw JSON (this is the test input for instance). The JSON is loaded into `PipelineState`, then each of the new functions are invoked for that state. This way it will (hopefully) be super clear what the expected output of each should be, given a specific state.  Ultimately it's up to you how you want to do testing.